### ISSUE TITLE
A11y keyboard focus control in the Cards Component

### ISSUE THREAD

**Issue by DaveBest99** [angular/components#3684]
#### Bug, feature request, or proposal:
Bug - The Card component has no HTML or ARIA markup for screen reader users.

#### What is the expected behavior?
Pressing the Tab key should move focus to the Card component, and the md-card should have a keyboard supported navigation structure markup.

#### What is the current behavior?
The demo displays a basic Card. The Card has no focusable element, and there is no content structure to distinguish between Card sections.

#### What are the steps to reproduce?
See below.

Providing a Plunker (or similar) is the *best* way to get the team to see your issue.
Plunker template: https://goo.gl/DlHd6U

#### What is the use-case or motivation for changing an existing behavior?
Creating a good keyboard user experience.

#### Which versions of Angular, Material, OS, browsers are affected?
Using screen reader JAWS17 with browsers Firefox, Internet Explorer 11, and Google Chrome, on Windows10.

#### Is there anything else we should know?
Screen reader user experience follows:

1. Press Tab key to the Card demo component in the Tab Panel.
JAWS says "like button", and the braille display shows "LIKE btn".
The screen reader focus skips over the Card Header and Content, and moves directly to the "LIKE" Button in the Card footer. The md-card object could use the tabindex=0 to include it in the focus order.
A card is a sheet of material that serves as an entry point to more detailed information. Cards display content composed of different elements whose size or supported actions vary, and can be displayed as a collection of Cards.
The Card Component supports gestures like Swipe and Pick-up-and-move, but does not define keyboard controls, and is structured with a hierarchy of Header, Content, and footer Actions that use no ARIA markup.

2. Use the arrow keys to move screen reader focus to the Tab Panel Card component.
JAWS says and displays "Shiba InuDog Breed".
As the user presses Down Arrow JAWS speaks only the line text, but does not indicate which lines belong to which sections of the Card, and does not identify the element type. All Card sections and elemetns are merged into a single block of text. This title, and the following paragraph, appear as just text with no context relationship.

3. The author must manage focus order for screen readers to understand content context.
- The Md-card module, or md-card-collection, should use tabindex=0, so as to be in the page focus order. Note, if there is a collection of Cards, they should be grouped into a container user role=group or role=grid properties, where each Card when activated opens a Modal Dialog.
- The md-card could be defined as a ARIA region, but since it is a container that has its own hierarchy structured with multiple elements, it would be better defined as an ARIA Dialog Modal.
- The md-card, and each Md-card section, should be defined as a ARIA Landmark region with an appropriate text label. Note, in this demo the md-card-header, md-card-content and md-card-footer are not marked up as defined Card sections.
- The Card modules, md-card-title, md-card-subtitle, and md-card-action	could be defined as an ARIA role=heading and the md-card could have role=banner if in a Dialog modal.
- The md-card-content could use role=main and md-card-footer could use role=contentinfo, with appropriate text labels, if in a Dialog Modal.

4. Move the screen reader focus to the Card Action buttons.
JAWS says "LIKE button" or "SHARE button".
Pressing Spacebar or Enter key on these buttons perform no action. This may be a feature of the demo Tab Panel or a defect in the md-card-action module.

See:
The Roles Model | Accessible Rich Internet Applications (WAI-ARIA) 1.0
http://w3c.github.io/aria-practices/#grid

WAI-ARIA Authoring Practices 1.1
2.9 Dialog (Modal)
http://w3c.github.io/aria-practices/#h-dialog_modal

Angular Material - Demos > Card
Note, it is impossible for a screen reader user to navigate and understand the Card content context in this demo.
https://material.angularjs.org/latest/demo/card

Demo - Angular Material - Component Card
https://material.angular.io/components/component/card

**Comment by jelbourn** [comment]
This is an interesting problem, since we had largely been considering the card to be a purely decorative treatment over otherwise normal text content. However, this doesn't provide as much context for screen-reader users because there's no real context for "this is a card". The `<figure>` element may be a good candidate for the root card content, but it's not clear which part of the content would be the `<figcaption>`. It would also be worth exploring whether `role="group"` would make a positive difference here. 

I suspect that, in the underlying component, there's a limited amount of information we can provide and that ultimately guidance documentation is necessary.

**Comment by DaveBest99** [comment]
The Card is a self-contained component (with Heading, Body, Footer) and must be defined as a block of content to the screen reader. That is, it must be distinguishable from the surrounding page content. The design of the Card should be carefully considered, since there are different types of Cards. The Card content could be enclosed in a landmark region (role=region aria-label="Card name"). However, the Headings within the Card should not break the Heading hierarchy of the surrounding page content. Better yet, it may be better to contain the Card content in a Modal Dialog, so that the keyboard user cannot move focus outside of the Card without closing it first. The WCAG technique should be consistent to establish the Angular brand.

 

David
